crates/red_knot_python_semantic/src/types/call/arguments.rs

```diff
@@ -98,18 +95,21 @@ impl<'a, 'db> CallArgumentTypes<'a, 'db> {
     /// function is invoked with the unmodified argument list.)
     pub(crate) fn with_self<F, R>(&self, bound_self: Option<Type<'db>>, f: F) -> R
     where
-        F: FnOnce(Self) -> R,
+        F: FnOnce(&Self) -> R,
```

Comment:
Could this return `CallArguments` now instead of taking a closure?

Reply:
Hmmm...almost. The wrinkle is that I'm trying to avoid cloning if there's no `bound_self`. So I think it could return a `Cow<CallArguments>`.

Reply:
Done